File: MVCProject/app/src/main/java/com/example/mvc_project/allproducts/controller/AllProductsActivity.java
package com.example.mvc_project.allproducts.controller;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.widget.Toast;

import com.example.mvc_project.R;
import com.example.mvc_project.allproducts.view.allAdabter;
import com.example.mvc_project.allproducts.view.onAdd;
import com.example.mvc_project.model.Product;
import com.example.mvc_project.model.Repository;
import com.example.mvc_project.network.NetworkDelegate;
import com.example.mvc_project.network.Product_Client;

import java.util.ArrayList;

public class AllProductsActivity extends AppCompatActivity implements NetworkDelegate , onAdd {

    RecyclerView recyclerView ;
    com.example.mvc_project.allproducts.view.allAdabter allAdabter;
    RecyclerView.LayoutManager layoutManager;
    Product_Client product_client ;

    onAdd onadd;

    Repository repo;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_all_products);
        recyclerView = findViewById(R.id.recycler_view);
        product_client =Product_Client.getInstance();
        product_client.enqueueCall(this);
        layoutManager = new LinearLayoutManager(this);
        repo = Repository.getInstance(this);

    }

    @Override
    public void onSuccessResponse(ArrayList<Product> products) {
        allAdabter = new allAdabter(this,products , this);
        recyclerView.setAdapter(allAdabter);
        recyclerView.setLayoutManager(layoutManager);
    }

    @Override
    public void onFailureResponse(String error) {

    }

    @Override
    public void onClick(Product product) {
            repo.insert(product);
            Toast.makeText(AllProductsActivity.this, "Added To Favourite", Toast.LENGTH_SHORT).show();
        }
    }

File: MVVM/app/src/main/java/com/example/mvvm/allProducts/view/AllProductsActivity.kt
package com.example.mvvm.allProducts.view

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.mvvm.R
import com.example.mvvm.allProducts.viewmodel.AllProductsViewModel
import com.example.mvvm.allProducts.viewmodel.AllProductsViewModelFactory
import com.example.mvvm.database.ConcretLocalSource
import com.example.mvvm.model.Product
import com.example.mvvm.model.Repository
import com.example.mvvm.network.ProductClient

class AllProductsActivity : AppCompatActivity() , OnClick{
    lateinit var allFactory:AllProductsViewModelFactory
    lateinit var viewModel:AllProductsViewModel
    lateinit var allRecycler:RecyclerView
    lateinit var allAdapter: AllAdapter
    lateinit var mylayoutManager: LinearLayoutManager
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_all_products)
        allRecycler = findViewById(R.id.all_rv)
        allAdapter  = AllAdapter(listOf(),this,this)
        mylayoutManager = LinearLayoutManager(this)
        allRecycler.layoutManager = mylayoutManager
        allRecycler.adapter = allAdapter
        allFactory = AllProductsViewModelFactory(Repository.getInstance(ProductClient.getInstance() ,ConcretLocalSource.getInstance(this)))
        viewModel = ViewModelProvider(this,allFactory).get(AllProductsViewModel::class.java)
        viewModel._myProducts.observe(this){products->

            if (products!=null){
                allAdapter.proudcts = products
                allAdapter.notifyDataSetChanged()
                Log.i("Milad",""+products.get(0))
                System.out.println(" MILAD SOLIMAN"+products.get(0))
            }
        }
    }


    override fun addToFavurote(product: Product) {
      viewModel.addProduct(product)
      Toast.makeText(this,"Added To Favourite",Toast.LENGTH_SHORT).show()
    }


}